components/org.wso2.carbon.aws/src/main/java/org/wso2/carbon/aws/user/store/mgt/util/AWSSignatureV4Generator.java

```diff
@@ -0,0 +1,424 @@
+                String key = entrySet.getKey();
+                String value = entrySet.getValue();
+                signedHeaders.append(key).append(";");
+                canonicalURL.append(key).append(":").append(value).append("\n");
+            }
+
+            /* Note: Each individual header is followed by a newline character, meaning the complete list ends with
+             a newline character. */
+            canonicalURL.append("\n");
+        } else {
+            canonicalURL.append("\n");
+        }
+
+        // Step 1.5 Add the signed headers, followed by a newline character.
+        signedHeaderString = signedHeaders.substring(0, signedHeaders.length() - 1); // Remove last ";"
+        canonicalURL.append(signedHeaderString).append("\n");
+
+        /* Step 1.6 Use a hash (digest) function like SHA256 to create a hashed value from the payload in the body of
+        the HTTP or HTTPS. */
+        if (payload == null) {
+            payload = "";
+        }
+        canonicalURL.append(hash(payload));
+        if (log.isDebugEnabled()) {
+            log.debug("Canonical Request: " + canonicalURL.toString());
+        }
+
+        return canonicalURL.toString();
+    }
+
+    /**
+     * Task 2: Create a String to Sign for Signature Version 4.
+     */
+    private String prepareStringToSign(String canonicalURL) {
+
+        String stringToSign;
+
+        // Step 2.1 Start with the algorithm designation, followed by a newline character.
+        stringToSign = AWSConstants.HMAC_ALGORITHM + "\n";
+
+        // Step 2.2 Append the request date value, followed by a newline character.
+        stringToSign += xAmzDate + "\n";
+
+        // Step 2.3 Append the credential scope value, followed by a newline character.
+        stringToSign += currentDate + "/" + regionName + "/" + serviceName + "/" + AWSConstants.AWS4_REQUEST + "\n";
+
+        /* Step 2.4 Append the hash of the canonical request that you created in Task 1: Create a Canonical Request
+        for Signature Version 4. */
+        stringToSign += hash(canonicalURL);
+
+        if (log.isDebugEnabled()) {
+            log.debug("String to sign: " + stringToSign);
+        }
+
+        return stringToSign;
+    }
+
+    /**
+     * Task 3: Calculate the AWS Signature Version 4.
+     */
+    private String calculateSignature(String stringToSign) {
+
+        try {
+            // Step 3.1 Derive your signing key.
+            byte[] signatureKey = getSignatureKey(secretAccessKey, currentDate, regionName, serviceName);
+
+            // Step 3.2 Calculate the signature.
+            byte[] signature = hmacSHA256(signatureKey, stringToSign);
+
+            // Step 3.2.1 Encode signature (byte[]) to Hex.
+            return bytesToHex(signature);
+        } catch (UnsupportedEncodingException | InvalidKeyException | NoSuchAlgorithmException ex) {
+            log.error("Error while calculating the signature." + ex);
```

Comment:
Why can't we simply wrap or re-throw this exception ? Do we need to handle this and return a null here ?